libexec/python/shell.py

```diff
@@ -99,63 +187,66 @@ def parse_image_uri(image,
+
+    # Be absolutely sure there are no comments
     image = image.split('#')[0]
 
-    # Get rid of any uri, and split the tag
-    image = image.replace(uri, '')
-
-    # Does the uri have a digest or Github tag (version)?
-    image = image.split('@')
-    version = None
-    if len(image) == 2:
-        version = image[1]
-
-    image = image[0]
-    image = image.split(':')
-
-    # If there are three parts, we have port and tag
-    if len(image) == 3:
-        repo_tag = image[2]
-        image = "%s:%s" % (image[0], image[1])
-
-    # If there are two parts, we have port or tag
-    elif len(image) == 2:
-        # If there isn't a slash in second part, we have a tag
-        if image[1].find("/") == -1:
-            repo_tag = image[1]
-            image = image[0]
-        # Otherwise we have a port and we merge the path
-        else:
-            image = "%s:%s" % (image[0], image[1])
-            repo_tag = default_tag
-    else:
-        image = image[0]
-        repo_tag = default_tag
+    if not uri:
+        uri = get_image_uri(image, quiet=True)
 
-    # Now look for registry, namespace, repo
-    image = image.split('/')
+    # docker images require slightly different rules
+    if uri == "docker://":
+        uri_regexes = [ _docker_uri_re ]
 
-    if len(image) > 2:
-        registry = image[0]
-        namespace = "/".join(image[1:-1])
-        repo_name = image[-1]
+    image = remove_image_uri(image, uri);
 
-    elif len(image) == 2:
-        registry = default_registry
-        namespace = image[0]
-        repo_name = image[1]
+    for r in uri_regexes:
+        match = r.match(image)
+        if match:
+            break
 
-    else:
+    if not match:
+        bot.error('Could not parse image "%s"! Exiting.' % image)
+        sys.exit(1)
+
+    registry = match.group('registry')
+    namespace = match.group('namespace')
+    repo_name = match.group('repo')
+    repo_tag = match.group('tag')
+    version = match.group('version')
+
+    if namespace:
+        # strip trailing /
+        namespace = namespace.rstrip('/')
+
+    # repo_name is required and enforced by the re (in theory)
+    # if not provided, re should not match
+    assert(repo_name)
+
+    # replace empty fields with defaults
+    if not namespace:
+        # for docker, if a registry was specified, don't
+        # inject a namespace in between
+        if registry and uri == "docker://":
+            namespace = ""
+        else:
+            namespace = default_namespace
+    if not registry:
         registry = default_registry
-        namespace = default_namespace
-        repo_name = image[0]
+    if not repo_tag:
+        repo_tag = default_tag
+    # version is not mandatory, don't check that
 
-    # if user gave custom registry / namespace, use
-    if CUSTOM_NAMESPACE is not None:
-        namespace = CUSTOM_NAMESPACE
-    if CUSTOM_REGISTRY is not None:
-        registry = CUSTOM_REGISTRY
 
     if not quiet:
```

Comment:
It looks like if the version is not long / missing numbers it gets parsed as part of the tag instead of version:

```
parse_image_uri('docker://localhost:5000/long/name/ubuntu:latest@2.2')
Out[46]: 
{'namespace': 'long/name',
 'registry': 'localhost:5000',
 'repo_name': 'ubuntu',
 'repo_tag': 'latest@2.2',
 'version': None}
```

this is ok for docker because we won't see much outside of sha256xxxxx, but for shub, a digest can also be a commit or a branch. For example, here we are seeing the digest added to the tag:

```
parse_image_uri('shub://namespace:5000/one/two/ubuntu:latest@master')

{'namespace': 'one/two',
 'registry': 'namespace:5000',
 'repo_name': 'ubuntu',
 'repo_tag': 'latest@master',
 'version': None}
```

**returns from 40 minutes of testing**
ok, so the slight mis-parsing isn't an issue with pulling images from shub because the url squashes the tag with the digest anyway, so it's just a designation for the function. In terms of function, this should be ok because it gets the job done. I'm more concerned with users seeing the mis-parsed uri and reporting issue, and it looks like it's set to be quiet so this won't come up. I think this is acceptable for now and we can revisit if needed.

Reply:
No, the mis-parsing is definitely an issue, good catch! That's the kind that if you ignore it now, you'll forget about it a year from now, then you'll spend hours (or days) debugging once you add a new feature which requires the `version` to be parsed correctly...
The fix is simple enough, I need to enforce, in the regex, that the tag cannot contain a `@`.

After locally fixing the above, I also found a couple of other interesting cases, which raise these questions:
1. Can `version` contain a `@` character? This is what happens with my code, after the fix:
```python
In [2]: parse_image_uri("docker://something.com/image:tag@2.2@3")
ERROR Could not parse image "something.com/image:tag@2.2@3"! Exiting.
An exception has occurred, use %tb to see the full traceback.

SystemExit: 1
```
2. Can `registry` contain a `@` character?
```python
In [6]: parse_image_uri("docker://user@something.com/image:tag@version")
Out[6]: 
{'namespace': '',
 'registry': 'user@something.com',
 'repo_name': 'image',
 'repo_tag': 'tag',
 'version': 'version'}
```
Should that be parsed as
```python
{'namespace': 'library',
 'registry': 'index.docker.io',
 'repo_name': 'user',
 'repo_tag': None,
 'version': 'something.com/image:tag@version'}
```
instead?
Of course, I kinda biased that example to make it look like `registry` should be as it is parsed by my code, but here's another example:
```python
In [16]: parse_image_uri("shub://ubuntu@some.version/something")
Out[16]: 
{'namespace': 'library',
 'registry': 'ubuntu@some.version',
 'repo_name': 'something',
 'repo_tag': 'latest',
 'version': None}
```

These are both easy to fix in one direction or another, just asking which way to go.

Reply:
As far as I know, I haven't seen a registry with an `@` character, so let's disregard that for now. As for the version, for singularity hub, this could be the case if the user has a branch with an `@` in it. It's likely a VERY rare case. To answer your question:

```
parse_image_uri("docker://something.com/image:tag@version")
```

should come out to be (note that I've removed the `user@` - have you seen this anywhere?

```
{'namespace': '',
 'registry': 'something.com',
 'repo_name': 'image',
 'repo_tag': 'tag',
 'version': 'version'}
```

Reply:
The `user@domain` syntax is used for e.g. HTTP Basic Auth, although docker by default uses a different authentication scheme, and itself doesn't accept URIs in that format.
I've found at least [one blog post](https://ahmet.im/blog/docker-http-basic-auth/) where somebody describe doing this, and a [related docker bug](https://github.com/moby/moby/issues/8283) (mentioned in that blog post) was closed with "Probably not so much needed", so if docker doesn't support that, I guess singularity doesn't need to support it either (yet?).

So, based on this, I'll change it to allow `@` inside version, and disallow it inside `registry`.